(Publish 15 May 2024 at 9:00 AM) Logs: Add query time parsing rules (#17276)
newrelic/docs-website: src/content/docs/logs/ui-data/query-time-parsing.mdx

            </td>
        </tr>
        <tr>
            <td>
                Timing
            </td>
            <td>
                Applied at ingest
            </td>
            <td>
                Applied when you query
            </td>
        </tr>
        <tr>
            <td>
                Results
            </td>
            <td>
                Makes permanent changes to stored log 
            </td>
            <td>
                Temporarily alters the logs table in the UI
            </td>
        </tr>
        <tr>
            <td>
                Live tail logs
            </td>
            <td>
                Live tail logs include any extracted values from ingest parsing
            </td>
            <td>
                Live tail logs **don't** include any extracted values from query time parsing
            </td>
        </tr>
        <tr>
            <td>
                Exported logs
            </td>
            <td>
                Exported logs include any extracted values from ingest parsing
            </td>
            <td>
                Exported logs **don't** include any extracted values from query time parsing

>             </td>

Reviewer: missing a closing `</tr>` here

import logsOpenQueryTimeParsingEditor from 'images/logs_screenshot-full_open-query-time-parsing-editor.webp'

import logsExtractValuesinParserRule from 'images/logs_screenshot-crop_extract-values-in-parser-rule.webp'

import logsEditorShowsParsedDataPreview from 'images/logs_screenshot-crop_editor-shows-parsed-data-preview.webp'

import logsNrqlBehindQueryRule from 'images/logs_screenshot-crop_nrql-behind-query-rule.webp'

import logsDeleteQueryTimeRule from 'images/logs_screenshot-crop_delete-query-time-rule.webp'

import logsViewResultsinLogDetailsView from 'images/logs_screenshot-full_view-results-in-log-details-view.webp'

import logsLogTableResults from 'images/logs_screenshot-full_log-table-results.webp'

import logsRuleManager from 'images/logs_screenshot-crop_rule-manager.webp'

import logsQueryTimeSearchBar from 'images/logs_screenshot-full_query-time-search-bar.webp'

import logsAlertQueryBuilder from 'images/logs_screenshot-crop_alert-query-builder.webp'

import logsQueryBehindParsingRule from 'images/logs_screenshot-crop_query-behind-parsing-rule.webp'

import logsParsingAttributesView from 'images/logs_screenshot-crop_parsing-attributes-view.webp'

> Query time parsing lets you parse your logs directly in the UI without needing to write complex regular expressions or grok patterns. Results are shown instantly since parsing is performed at query time. You can use query time parsing to temporarily extract values from your logs and quickly perform a query on these variables.

Reviewer 2: The word Grok here feels out of place. Primarily because it feels like slang or a colloquial term, but also I think it gets sticky with New Relic Grok. I'd love it if you could get more specific, which might be hard to do for the first sentence. But why does query time parsing mean you don't have to use regular expressions?

Reviewer 2: I revisited this- I think I was misunderstanding Grok. We can sync about this later. I still think there needs to be a greater call to action at the beginning. I think a use case or an example would be super helpful here. It would also be in line with what our customers request. It feels like I don't have anything to ground myself within. Why is this a special new feature? Why use this rather than another method other than to save yourself time?

Author: I've revised the introduction.

## How query time parsing differs from ingest parsing [#differences]

While both types of parsing make it easier for you to query logs, they have some significant differences:

> **Ingest parsing:** Parsing during [log ingestion](/docs/logs/ui-data/parsing/) is where you create parsing rules using grok and/or regex. As log records are ingested at New Relic, your parsing rules are applied to create permanent attributes that are stored with your log data in NRDB. These attributes make it easier for you to query log data. 

Reviewer 2: Can you indent these or bullet them? Right now, since they're left-justified, they don't feel like you're listing two significant differences. They should stand out more visually

Author: @homelessbirds good suggestion! I will indent these!

**Query time parsing:** In contrast to ingest parsing, query time parsing lets you create temporary attributes that will be used as query variables. You can then use these variables in NRQL queries to populate your log table. We automatically create the queries as you make selections in the UI. 

> You may choose to use a combination of both parsing approaches. Review the table below to decide if query time parsing is right for you:

Reviewer 2: Should it be "You may also choose..."

Author: @homelessbirds good suggestion!

<table>
    <thead>
        <tr>
            <th>
                Description 
            </th>
            <th>
                Ingest parsing 
            </th>
            <th>
                Query time parsing 
            </th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>
                Recommended usage
            </td>
            <td>
                Best for creating permanent attributes you can query in the future
            </td>
            <td>
                Best for doing quick queries on attributes that aren't permanent 
            </td>
        </tr>
        <tr>
            <td>
                Parsing language                
            </td>
            <td>

>                 You create grok patterns and regular expressions 

Reviewer 2: Maybe I'm confused actually! What is Grok? Are you speaking about something other than New Relic AI? I've researched it a little and all I am coming up with is Elon Musk's Grok? Is there something I am missing?

Author: @homelessbirds I'm going to capitalize Grok since we do that in other docs. It's a special query language for logs that's a dialect of regular expressions.

            </td>
            <td>
                New Relic creates queries for you using the NRQL [aparse](/docs/nrql/nrql-syntax-clauses-functions/#func-aparse) function 
            </td>
        </tr>
        <tr>
            <td>
                Timing
            </td>
            <td>
                Applied at ingest
            </td>
            <td>
                Applied when you query
            </td>
        </tr>
        <tr>
            <td>
                Results
            </td>
            <td>
                Makes permanent changes to stored log 
            </td>
            <td>

>                 Temporarily alters the logs table in the UI

Reviewer 3: Could we change the phrasing of the alteration to target logs instead of the logs table? I think it would be more precise to say that query time parsing affects the logs themselves. Something similar to this would be great:
`Temporarily alters your logs in the UI`

Author: @mchavez-newrelic yes, I will push up this change shortly.

                Live tail logs
            </td>
            <td>
                Live tail logs include any extracted values from ingest parsing
            </td>
            <td>
                Live tail logs **don't** include any extracted values from query time parsing
            </td>
        </tr>
        <tr>
            <td>
                Exported logs
            </td>
            <td>
                Exported logs include any extracted values from ingest parsing
            </td>
            <td>
                Exported logs **don't** include any extracted values from query time parsing
            </td>
        </tr>
        <tr>
            <td>
                Number of attributes
            </td>
            <td>
                A maximum of [255 attributes](/docs/logs/log-api/log-event-data/#events) is available at ingest (the actual number attributes you can parse at ingest depends on the nature of your logs)
            </td>
            <td>
                You can parse a maximum of 32 temporary attributes across all rules for query time parsing
            </td>
        </tr>
    </tbody>
</table>

## How to create a query time parsing rule [#how-to] 

> Here's an overview of the tasks for creating query time parsing rules:

Reviewer 2: What about, "Here's a guide to create query time parsing rules"- tasks and guide have different tones.

Author: @homelessbirds I like this! I'll use it!

<Steps>
    <Step>
        #### Select attribute value to parse [#select-attribute]

>         You can start creating a query time parsing rule by selecting an attribute to parse:

Reviewer 3: Could we say "attribute value" instead of "attribute" to parse?

Author: @mchavez-newrelic yes, I will push up this change shortly.

            1. In the log table or in the **Formatted** tab of the log details view, highlight an attribute value you'd like to parse, known as the anchor string. Keep the following in mind:
                * Your highlighted text should include the value(s) you want to extract and the surrounding string characters that will help identify the location of the extracted value(s). 

>                 * If your initial anchor string is anywhere in the middle of the original attribute value, include at least one character before and after the values you want to extract.

Reviewer 2: Can your SME's provide you with a sample? It would be nice to be able to see the anchor string you're talking about/see it broken up/in action

Author: @homelessbirds I'll see if I can add the example from the screenshot right into the procedure wording.

                * If you're highlighting the entire attribute value, you don't need to worry about characters before and after the values you're extracting.
        <Callout variant="tip">
            You cannot highlight blob values to parse.
        </Callout>
            2. Click the **Create query time parsing rule** option. 

                **Log table:**

>                 <img

Reviewer 2: Can you add an arrow to this visual? The green box blends so well into the UI it's difficult to see.

Author: @homelessbirds yes, I will add a green arrow from the log record to the menu.

                    title="Screenshot showing how you can begin with the logs table UI"
                    alt="Screenshot showing how you can begin with the logs table UI"
                    src={logsLogsTableEntryPoint}
                />

                **Log details view**:
                <img
                    title="Screenshot showing how you can start creating rules in the logs detail view"
                    alt="Screenshot showing how you can start creating rules in the logs detail view."
                    src={logsLogsDetailEntryPoint}
                />
    </Step>
    <Step>
        #### Highlight and extract values [#highlight-extractl]

        After you've clicked **Create query time parsing rule**, the editor displays the string you selected for parsing:

        <img

>         title="Screenshot showing the opening screen of the query editor"

Reviewer 2: I see how you have the examples displayed in the screenshot but I still wonder if it would be more effective/helpful to have examples brought into the text

        alt="Screenshot showing the opening screen of the query editor"
        src={logsOpenQueryTimeParsingEditor}
        />

        To extract values:

        1. Within the string, highlight the value(s) you want to extract (see [Tips for extracting values](#extract-tips)).
        2. Under **Parse as**, enter a name for this temporary attribute that will be used as a query variable. 
            <img
            title="Screenshot showing how to extract values in the parser rule"
            alt="Screenshot showing how to extract values in the parser rule"
            src={logsExtractValuesinParserRule}
            />
        3. Click **Save**, which replaces the value you highlighted with the variable you created.

> <CollapserGroup>

Reviewer 2: This is a great use of a collapser. Really well done.

    </Collapser>
</CollapserGroup>

    </Step>
    <Step>
        #### Finish creating your query time parsing rule [#finish-rule]

        After you've selected values, complete the following:

        1. In the editor, review the preview of the table showing how your rule will be applied to the log you selected. 

            <img
                title="Screenshot showing how the editor shows a preview of the rule"
                alt="Screenshot showing how the editor shows a preview of the rule"
                src={logsEditorShowsParsedDataPreview}
            />

        2. If you're interested in the NRQL function used in the query to get your logs, click on **Query**. 

            <img
                title="Screenshot showing the query behind the parsing rule"
                alt="Screenshot showng the query behind the parsing rule"
                src={logsNrqlBehindQueryRule}
            />

            <figcaption>
            The pattern string, which is used to find and extract value(s) from the selected attribute, is updated whenever a value is extracted.
            </figcaption>

>         3. If you need to rename any variables before you create your rule, click on the name, provide a new name, and the click **Save**. You can also delete variables by clicking on the variable you wish to delete and clicking **Delete**.

Reviewer 3: Very small suggestion, should this be the word `then`?
```suggestion
        3. If you need to rename any variables before you create your rule, click on the name, provide a new name, and then click **Save**. You can also delete variables by clicking on the variable you wish to delete and clicking **Delete**.
```

            <img
                title="Screenshot showing where you can delete values"
                alt="Screenshot showing where you can delete values"
                src={logsDeleteQueryTimeRule}
            />
            <figcaption>
            You can edit or delete variables by clicking on them.
            </figcaption>
        4. After you've finished reviewing and editing your variable names, click **Create rule** to finish creating your rule. 

>         5. If you started creating a query time parsing rule from the **Log details** view, you need to close that view to see your results in the log table.

Reviewer 2: Is this a step or is this just a tip or a note?

Author: @homelessbirds you're right that this wasn't really a step. It's more of a note. I moved this into a paragraph outside the procedure. What do you think?

                src={logsViewResultsinLogDetailsView}
            />

            <figcaption>
            After you view your newly extracted values, close the **Log details** view. 
            </figcaption>

            **Log table**
            <img
                title="Screenshot showing the final results after you create the rules"
                alt="Screenshot showing the final results after you create the rules"
                src={logsLogTableResults}
            />

            <figcaption>
                The log table automatically refreshes and applies the generated query to parse your logs.
            </figcaption>
    </Step>
</Steps>

## Manage your rules [#manage-rules]

While your rules are temporary and apply to your current user session, you can still perform a variety of tasks during your session.

<CollapserGroup>
    <Collapser
        id="edit-and-delete-rules"
        title="Edit and delete rules"
    >
      You can edit or delete your rules by clicking **Manage parsing rules** above the log table. These rules are stored in your current user session within the logs view. 

>       <img

Author: @homelessbirds I added a green arrow to the three dots menu.

import logsOpenQueryTimeParsingEditor from 'images/logs_screenshot-full_open-query-time-parsing-editor.webp'

import logsExtractValuesinParserRule from 'images/logs_screenshot-crop_extract-values-in-parser-rule.webp'

import logsEditorShowsParsedDataPreview from 'images/logs_screenshot-crop_editor-shows-parsed-data-preview.webp'

import logsNrqlBehindQueryRule from 'images/logs_screenshot-crop_nrql-behind-query-rule.webp'

import logsDeleteQueryTimeRule from 'images/logs_screenshot-crop_delete-query-time-rule.webp'

import logsViewResultsinLogDetailsView from 'images/logs_screenshot-full_view-results-in-log-details-view.webp'

import logsLogTableResults from 'images/logs_screenshot-full_log-table-results.webp'

import logsRuleManager from 'images/logs_screenshot-crop_rule-manager.webp'

import logsQueryTimeSearchBar from 'images/logs_screenshot-full_query-time-search-bar.webp'

import logsAlertQueryBuilder from 'images/logs_screenshot-crop_alert-query-builder.webp'

import logsQueryBehindParsingRule from 'images/logs_screenshot-crop_query-behind-parsing-rule.webp'

import logsParsingAttributesView from 'images/logs_screenshot-crop_parsing-attributes-view.webp'

> Are you looking for a quick way to list log records that contain certain strings? Query time parsing lets you parse your logs directly in the UI without needing to write complex regular expressions or Grok patterns. You can use query time parsing to temporarily extract values from your logs and quickly perform a query on these variables. The results are shown instantly since parsing is performed at query time. 

Reviewer 3: I like having the the call to action at the beginning! If possible, I'd like to suggest a slightly modified CTA that highlights the differences from ingest parsing. Let me know your thoughts. I'm also open to other suggestions!

```
Are you looking for a quick way to visually extract attributes from your logs after they've been ingested into New Relic?
```

Author: @mchavez-newrelic this is great! I'll use your sentence!

## How query time parsing differs from ingest parsing [#differences]

While both types of parsing make it easier for you to query logs, they have some significant differences:

> * **Ingest parsing:** Parsing during [log ingestion](/docs/logs/ui-data/parsing/) is where you create parsing rules using grok and/or regex. As log records are ingested at New Relic, your parsing rules are applied to create permanent attributes that are stored with your log data in NRDB. These attributes make it easier for you to query log data. 

Reviewer 3: Small suggestion for capitalizing grok here.
```suggestion
* **Ingest parsing:** Parsing during [log ingestion](/docs/logs/ui-data/parsing/) is where you create parsing rules using Grok and/or regex. As log records are ingested at New Relic, your parsing rules are applied to create permanent attributes that are stored with your log data in NRDB. These attributes make it easier for you to query log data. 
```

* **Query time parsing:** In contrast to ingest parsing, query time parsing lets you create temporary attributes that will be used as query variables. You can then use these variables in NRQL queries to populate your log table. We automatically create the queries as you make selections in the UI. 

You may also choose to use a combination of both parsing approaches. Review the table below to decide if query time parsing is right for you:

<table>
    <thead>
        <tr>
            <th>
                Description 
            </th>
            <th>
                Ingest parsing 
            </th>
            <th>
                Query time parsing 
            </th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>
                Recommended usage
            </td>
            <td>
                Best for creating permanent attributes you can query in the future
            </td>
            <td>
                Best for doing quick queries on attributes that aren't permanent 
            </td>
        </tr>
        <tr>
            <td>
                Parsing language                
            </td>
            <td>

>                 You create grok patterns and regular expressions 

Reviewer 3: Ah sorry I missed this instance of capitalizing Grok as well. I believe this is the last one.

Author: @mchavez-newrelic another good catch! I'll fix this!

                Live tail logs
            </td>
            <td>
                Live tail logs include any extracted values from ingest parsing
            </td>
            <td>
                Live tail logs **don't** include any extracted values from query time parsing
            </td>
        </tr>
        <tr>
            <td>
                Exported logs
            </td>
            <td>
                Exported logs include any extracted values from ingest parsing
            </td>
            <td>
                Exported logs **don't** include any extracted values from query time parsing
            </td>
        </tr>
        <tr>
            <td>
                Number of attributes
            </td>
            <td>
                A maximum of [255 attributes](/docs/logs/log-api/log-event-data/#events) is available at ingest (the actual number attributes you can parse at ingest depends on the nature of your logs)
            </td>
            <td>
                You can parse a maximum of 32 temporary attributes across all rules for query time parsing
            </td>
        </tr>
    </tbody>
</table>

## How to create a query time parsing rule [#how-to] 

> Here's a guide to creating query time parsing rules. The example shows how to list all the logs that contain the string `Running script`. 

Reviewer 3: If the call to action is modified in my above comment, I'd like to modify the wording here to align with that CTA. Let me know your thoughts!

```suggestion
Here's a guide to creating query time parsing rules. The example shows how to extract the log level and scripted message values from the message attribute. 
```

<Steps>
    <Step>
        #### Select attribute value to parse [#select-attribute]
        You can start creating a query time parsing rule by selecting an attribute value to parse.  

>             1. In the log table or in the **Formatted** tab of the log details view, highlight an anchor string that contains the string you want to extract. In this case, you'd highlight `level=info msg="Running script"`. It looks like this in the logs table:

Reviewer 3: Small suggestion, if we could use the term "values" for what's being extracted, that would be great.
```suggestion
            1. In the log table or in the **Formatted** tab of the log details view, highlight an anchor string that contains the values you want to extract. In this case, you'd highlight `level=info msg="Running script"`. It looks like this in the logs table:
```
